sdk/python/kfp/v2/dsl/component_bridge.py

```diff
@@ -136,15 +139,23 @@ def _resolve_output_path_placeholder(output_key: str) -> str:
     else:
       return _output_artifact_placeholder(output_key)
 
-  # IR placeholders are decided merely based on the declared type of the input.
-  # It doesn't matter wether it's InputValuePlaceholder or InputPathPlaceholder
-  # from component_spec.
-  placeholder_arguments = {
-      input_spec.name: _input_artifact_placeholder(input_spec.name)
-      if type_utils.is_artifact_type(input_spec.type) else
-      _input_parameter_placeholder(input_spec.name)
-      for input_spec in component_spec.inputs or []
-  }
+  placeholder_arguments = {}
+  for input_spec in component_spec.inputs or []:
+    # Skip inputs not presented in arguments list. This could happen for
+    # optional inputs whose presence in command/arguments is controlled by
+    # if placeholder.
+    if input_spec.name not in arguments:
+      continue
+
+    # IR placeholders are decided merely based on the declared type of the
+    # input. It doesn't matter wether it's InputValuePlaceholder or
```

Comment:
IMO at least this worths a warning: Value and Path/Uri are two drastically different concept. Abusing/overloading these two terms seem to be dangerous.

Reply:
This probably worth a separate discussion and PR if decided. 

In short, I have some concerns regarding whether such a warning is actionable -- like are we suggesting users to use {inputPath} when they want to specify a URI? Besides, there's consistence issue with output -- we only have {outputPath} placeholder today, so unless we introduce a new placeholder, we have the abusing/overloading problem anyway.

Reply:
IMHO I think it's not that hard to make a solid line between these two:

1. Path refers the uri/local path or whatever identifier that allows the user to access the data, while;

2. Value refers to the actual content of the data.

Semantically, for all the artifact types, `InputValue` placeholder is still meaningful. It is the decision of the system *not* to pass the actual value of the data to the command line interface. Instead we provide `InputPath` to let users access the data using whatever library they see fit. For output the story is a bit of different. The only thing available now is `OutputPath` because file-based output seems like the only container-friendly option at this moment. This is orthogonal to the artifact/parameter type ontology.

> like are we suggesting users to use {inputPath} when they want to specify a URI?

Yes. I think we should suggest them to use the canonical way to specify a URI, or local path if they're abusing the notation now.

Reply:
>> like are we suggesting users to use {inputPath} when they want to specify a URI?
>
> Yes. I think we should suggest them to use the canonical way to specify a URI, or local path if they're abusing the notation now.

This is a very special case, probably only for the type `GCSPath` given the legacy use pattern. I thought we should instead suggest users not to specify the URI at all, given now they don't have to do so? 
Anyway, this piece of logic isn't new to this PR, and not really related to if-placeholder. Let's sync offline, and maybe address it in a separate PR.

Reply:
Sounds good.

Reply:
I'd consider the following two cases valid:
* inputValue: Model **URI**
* inputPath: Model

Other combinations seem like invalid usages:
* inputValue: Model  # Trying to consume a binary (or multi-file) data by value
* inputPath: Model **URI**  # Expects file with a URI inside (not incorrect per se, but nobody does that)

The semantical contract for `inputPath` is that it should always resolve to a path that can be opened using `open()`. The `inputPath` placeholder **must not** ever be resolved to something with a schema (e.g. 'gs://...').

If the component code expects a URI, then this component should have a `Something URI` input and consume that input URI as a value using `inputValue`.

```
inputs:
- {name: Model URI, type: GcsPath}
implementation:
  container:
    image: ...
    command:
    - deploy
    - --model-uri
    - {inputValue: Model URI}
```